[CI] IndexRecoveryIT.testRecoverLocallyUpToGlobalCheckpoint fails intermittently

Got this error on CI today in the 7.7 branch. Looks like it has failed a few times in the last month or so on a few PR builds.

The failure:
```
org.elasticsearch.indices.recovery.IndexRecoveryIT > testRecoverLocallyUpToGlobalCheckpoint FAILED
 java.lang.AssertionError: total recovered translog operations must include both local and remote recovery
 Expected: a value equal to or greater than <89>
 but: <81> was less than <89>
 at __randomizedtesting.SeedInfo.seed([EFD4DFF3F294AE9E:158F52B7E5910AD6]:0)
 at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
 at org.junit.Assert.assertThat(Assert.java:956)
 at org.elasticsearch.indices.recovery.IndexRecoveryIT.testRecoverLocallyUpToGlobalCheckpoint(IndexRecoveryIT.java:1203)
REPRODUCE WITH: ./gradlew ':server:integTest' --tests "org.elasticsearch.indices.recovery.IndexRecoveryIT.testRecoverLocallyUpToGlobalCheckpoint" -Dtests.seed=EFD4DFF3F294AE9E -Dtests.security.manager=true -Dtests.locale=sq-AL -Dtests.timezone=Asia/Beirut -Dcompiler.java=13
```

Scan: https://gradle-enterprise.elastic.co/s/c5ubooujvkxoy

This failed again but with a different error this time:

```
java.lang.AssertionError: timed out waiting for green state
at org.junit.Assert.fail(Assert.java:88)
at org.elasticsearch.test.ESIntegTestCase.ensureColor(ESIntegTestCase.java:940)
at org.elasticsearch.test.ESIntegTestCase.ensureGreen(ESIntegTestCase.java:893)
at org.elasticsearch.test.ESIntegTestCase.ensureGreen(ESIntegTestCase.java:882)
at org.elasticsearch.indices.recovery.IndexRecoveryIT.testRecoverLocallyUpToGlobalCheckpoint(IndexRecoveryIT.java:1212)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
```

Jenkins: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+multijob-unix-compatibility/os=sles-12&&immutable/687/console
Build Scan: https://gradle-enterprise.elastic.co/s/fjxtmcsyeapuy

--- a/server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java
+++ b/server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java
@@ -20,6 +20,7 @@
 package org.elasticsearch.indices.recovery;
 
 import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.util.SetOnce;
 import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.Version;
 import org.elasticsearch.action.ActionRunnable;
@@ -1135,12 +1136,25 @@ public class IndexRecoveryIT extends ESIntegTestCase {
         client().admin().indices().prepareRefresh(indexName).get(); // avoid refresh when we are failing a shard
         String failingNode = randomFrom(nodes);
         PlainActionFuture<StartRecoveryRequest> startRecoveryRequestFuture = new PlainActionFuture<>();
+        // Peer recovery fails if the primary does not see the recovering replica in the replication group (when the cluster state
+        // update on the primary is delayed). To verify the local recovery stats, we have to manually remember this value in the
+        // first try because the local recovery happens once and its stats is reset when the recovery fails.
+        SetOnce<Integer> localRecoveredOps = new SetOnce<>();
         for (String node : nodes) {
             MockTransportService transportService = (MockTransportService) internalCluster().getInstance(TransportService.class, node);
             transportService.addSendBehavior((connection, requestId, action, request, options) -> {
                 if (action.equals(PeerRecoverySourceService.Actions.START_RECOVERY)) {
-                    assertFalse("recovery request was set already", startRecoveryRequestFuture.isDone());
-                    startRecoveryRequestFuture.onResponse((StartRecoveryRequest) request);
+                    final RecoveryState recoveryState = internalCluster().getInstance(IndicesService.class, failingNode)
+                        .getShardOrNull(new ShardId(resolveIndex(indexName), 0)).recoveryState();
+                    assertThat(recoveryState.getTranslog().recoveredOperations(), equalTo(recoveryState.getTranslog().totalLocal()));
+                    if (startRecoveryRequestFuture.isDone()) {
+                        assertThat(recoveryState.getTranslog().totalLocal(), equalTo(0));
+                        recoveryState.getTranslog().totalLocal(localRecoveredOps.get());
+                        recoveryState.getTranslog().incrementRecoveredOperations(localRecoveredOps.get());
+                    } else {
+                        localRecoveredOps.set(recoveryState.getTranslog().totalLocal());
+                        startRecoveryRequestFuture.onResponse((StartRecoveryRequest) request);
+                    }
                 }
                 if (action.equals(PeerRecoveryTargetService.Actions.FILE_CHUNK)) {
                     RetentionLeases retentionLeases = internalCluster().getInstance(IndicesService.class, node)
@@ -1171,6 +1185,7 @@ public class IndexRecoveryIT extends ESIntegTestCase {
         assertThat(commitInfoAfterLocalRecovery.maxSeqNo, equalTo(lastSyncedGlobalCheckpoint));
         assertThat(startRecoveryRequest.startingSeqNo(), equalTo(lastSyncedGlobalCheckpoint + 1));
         ensureGreen(indexName);
+        assertThat((long) localRecoveredOps.get(), equalTo(lastSyncedGlobalCheckpoint - localCheckpointOfSafeCommit));
         for (RecoveryState recoveryState : client().admin().indices().prepareRecoveries().get().shardRecoveryStates().get(indexName)) {
             if (startRecoveryRequest.targetNode().equals(recoveryState.getTargetNode())) {
                 assertThat("expect an operation-based recovery", recoveryState.getIndex().fileDetails(), empty());

